sycl/include/CL/sycl/INTEL/esimd/esimd_libcm.hpp

```diff
@@ -0,0 +1,311 @@
+    for (int I = 0; I < __Dims__; ++I) {
+      spaceDim[I] = (uint32_t)Range[I];
+    }
+    run();
+  }
+
+  // Item w/o offset
+  template <class ArgT = KernelArgType>
+  typename std::enable_if<
+      std::is_same<ArgT, sycl::item<__Dims__, /*Offset=*/false>>::value>::type
+  runIterationSpace(const sycl::range<__Dims__> Range) {
+    sycl::detail::NDLoop<__Dims__>::iterate(
+        Range, [&](const sycl::id<__Dims__> ID) {
+          sycl::item<__Dims__, /*Offset=*/false> Item =
+              IDBuilder::createItem<__Dims__, false>(Range, ID);
+          argVector.push_back(Item);
+        });
+    for (int I = 0; I < __Dims__; ++I) {
+      spaceDim[I] = (uint32_t)Range[I];
+    }
+    run();
+  }
+
+  // Item w/ offset
+  template <class ArgT = KernelArgType>
+  typename std::enable_if<
+      std::is_same<ArgT, sycl::item<__Dims__, /*Offset=*/true>>::value>::type
+  runIterationSpace(const sycl::range<__Dims__> Range,
+                    const sycl::id<__Dims__> Offset) {
+    sycl::detail::NDLoop<__Dims__>::iterate(
+        Range, [&](const sycl::id<__Dims__> &ID) {
+          sycl::id<__Dims__> OffsetID = ID + Offset;
+          sycl::item<__Dims__, /*Offset=*/true> Item =
+              IDBuilder::createItem<__Dims__, true>(Range, OffsetID, Offset);
+          argVector.push_back(Item);
+        });
+    for (int I = 0; I < __Dims__; ++I) {
+      spaceDim[I] = (uint32_t)Range[I];
+    }
+    run();
+  }
+
+  // NDItem
+  template <class ArgT = KernelArgType>
+  typename std::enable_if<
+      std::is_same<ArgT, sycl::nd_item<__Dims__>>::value>::type
+  runIterationSpace(const sycl::id<__Dims__> &GroupID,
+                    const sycl::group<__Dims__> Group,
+                    const sycl::range<__Dims__> LocalSize,
+                    const sycl::range<__Dims__> GlobalSize,
+                    const sycl::id<__Dims__> GlobalOffset) {
+    sycl::detail::NDLoop<__Dims__>::iterate(
+        LocalSize, [&](const sycl::id<__Dims__> &LocalID) {
+          sycl::id<__Dims__> GlobalID =
+              GroupID * LocalSize + LocalID + GlobalOffset;
+          const sycl::item<__Dims__, /*Offset=*/true> GlobalItem =
+              IDBuilder::createItem<__Dims__, true>(GlobalSize, GlobalID,
+                                                    GlobalOffset);
+          const sycl::item<__Dims__, /*Offset=*/false> LocalItem =
+              IDBuilder::createItem<__Dims__, false>(LocalSize, LocalID);
+          const sycl::nd_item<__Dims__> NDItem =
+              IDBuilder::createNDItem<__Dims__>(GlobalItem, LocalItem, Group);
+          argVector.push_back(NDItem);
+        });
+    for (int I = 0; I < __Dims__; ++I) {
+      spaceDim[I] = (uint32_t)LocalSize[I];
+    }
+    run();
+  }
+
+private:
+  template <class ArgT = KernelArgType>
+  typename std::enable_if<std::is_same<ArgT, sycl::id<1>>::value>::type run() {
+    struct LambdaWrapper_ID_1DIM *wrappedLambda_ID_1DIM =
```

Comment:
Could you please clarify where this structure if free'ed?

Reply:
I'll look up. If there is no 'free', I'll add one. If there is, I'll put comments.